### PR TITLE
Merge remappings

### PR DESCRIPTION
**Pull request by onbjerg**
If you create a `remappings.txt` file after the config then auto-generated remappings are not included, so you can't remap just one dependency. In other words the remappings behave like an "or" (if env is set use that, otherwise use remappings file, otherwise use CLI parameters, otherwise use auto-generated remappings)

This PR re-implements the behavior in https://github.com/gakonst/foundry/pull/519

### CODE REVIEW

**Comment by mattsse** [comment]
both solutions have drawbacks, if they stack, then you can't remove one, if they don't, then you always need to provide all,

I guess it comes down to what is most commonly used and then it makes sense to make them stack, I guess.

I implemented this previously, can you have a look at this:

https://github.com/gakonst/foundry/pull/297/commits/05607fdbcbd6c5f01c236ccf7aec33c2f3e4821d#diff-4429d8a75f4f230ae4b5c590d76a1a5a9aa1f2553bc4740c5f3d7dc0137cd704R767-R803


if we make them stack, then we also need to fix some test and need to decide what we do with the final figment layer, the cli `remapping` args, also stack? then we need to adjust that in the `BuildArgs` `Provider` implementation

**Comment by onbjerg** [comment]
You can remove remappings by specifying the same remapping name, e.g. if an autogenerated remapping is `@openzeppelin/=node_modules/@openzeppelin/openzeppelin-contracts` and you specify `@openzeppelin/=node_modules/@openzeppelin/openzeppelin-contracts/contracts` then the one you specified will overwrite the old one due to how the sort and dedup works

I agree both have drawbacks. I'm not sure which one is the most common, but have had at least one support ticket related to this so I think the non-stacking version might be counter-intuitive?

In terms of CLI arg stacking, I would say yeah. Essentially, anywhere you specify a remapping is an override with the env vars having final authority.

We should definitely discuss - thoughts @gakonst?

Not entirely sure, but I think dapptools does actually not merge

**Comment by gakonst** [comment]
Not sure I understand exactly the difference above or what "stack" means here?

If I have manually provided a remapping in remappings.txt, and then there's an auto-generated one for the same file, I'd prefer it to be less "restrictive" and allow using both (e..g in the OZ example above, if I've specified @openzeppelin/contracts and openzeppelin-contracts, somebody should be able to import in both ways). Maybe that's not ideal though and we should decide to only use the overridden version.

As a relevant point to remappings: Should we detect if a package.json exists in the remapping's directory, and instead of using the directory as the name of the remapping, use the package.json's name field? Or should we hardcode some remappings which are common e,g. `@openzeppelin/contracts` over `openzeppelin-cotnracts`?

**Comment by onbjerg** [comment]
Maybe stack is a bad choice of words, it's more like "merging". If you have two remappings with the same name from multiple sources, e.g.:

```sh
# Autogenerated remappings
my-lib/=lib/my-lib/
```

```
# My remappings.txt
my-lib/=lib/my-lib/src/
```

Then the final remapping would be `my-lib/=lib/my-lib/src/` since it discards remappings with the same name.

However, the current behavior since config PR was merged is an "or", along the lines of:

```
# Autogenerated remappings
my-lib/=lib/my-lib/
other-lib/=lib/other-lib/
```

```
# My remappings.txt
my-lib/=lib/my-lib/src/
```

The final remappings would just be

```
# My remappings.txt
my-lib/=lib/my-lib/src/
```

Since it completely overrides the auto-generated ones - i.e. there is no merging. This behavior is the same across the board - if you specify remappings in env vars, only those are used, if you specify it as CLI parameters, only those are used etc.

**Comment by gakonst** [comment]
OK so this PR's intention is to make it so that the final result in your example where there's 2 autogen'ed remappings + 1 override, into the following?
```
my-lib/=lib/my-lib/src/
other-lib/=lib/other-lib/
```

If so, that makes sense. We want to only _selectively_ override remappings, vs completely replacing them

**Comment by onbjerg** [comment]
Yes, that would be the final result if this PR is merged (after fixing the tests)

**Comment by gakonst** [comment]
Yeah supportive then, unless @mattsse has any strong disagreement. IMO overriding selectively makes more sense

File: config/src/lib.rs [hunk 1]
```diff
@@ -771,33 +771,43 @@ impl<'a> RemappingsProvider<'a> {
     /// - Environment variables
     /// - CLI parameters
     fn get_remappings(&self, remappings: Vec<Remapping>) -> Result<Vec<Remapping>, Error> {
```
Comment (onbjerg) [inline config/src/lib.rs:773]: I was unsure when to merge these in. Where are these remappings from? @mattsse

File: config/src/lib.rs [hunk 2]
```diff
@@ -771,33 +771,43 @@ impl<'a> RemappingsProvider<'a> {
...
-                .collect())
-        } else {
-            Ok(remappings)
-        }
+                .collect::<Vec<Remapping>>(),
+        );
+
+        // remove duplicates
+        new_remappings.sort_by(|a, b| a.name.cmp(&b.name));
+        new_remappings.dedup_by(|a, b| a.name.eq(&b.name));
```
Comment (mattsse) [inline config/src/lib.rs:808]: This works because `sort_by` does not reorder equal keys

**Review by mattsse — APPROVED** [review]

**Comment by mattsse** [comment]
There's one more thing to do

**Comment by gakonst** [comment]
nice @onbjerg @mattsse

**Pull request merged**